File: 02/solution.php
<?php
function lines(): Generator
{
    $file = fopen('input.txt', 'r');
    while (($line = fgets($file)) !== false) {
        yield trim($line);
    }
    fclose($file);
}

function plays($lines): Generator
{
    foreach($lines as $line) {
        [$opponent, $you] = explode(" ", $line);
        yield [ord($opponent) - ord("A"), ord($you) - ord("X")];
    }
}

function corrected_plays($plays): Generator
{
    // now map X, Y, Z onto lose/tie/win instead of rock/paper/scissors
    foreach($plays as [$opponent, $you]) {
        yield [$opponent, ($opponent + $you - 1 + 3) % 3];
    }
}

function scores($plays): Generator
{
    foreach($plays as $play) {
        list($opponent, $you) = $play;
        // PHP modulus operator returns same sign as first operand. To get
        // positive value out ensure we put positive value in.
        yield 1 + $you + (($you - $opponent + 1 + 3) % 3) * 3;
    }
}

function iterator_sum($iterator): int
{
    $total = 0;
    foreach($iterator as $item) {
        $total += $item;
    }
    return $total;
}

function solution1(): int
{
    return iterator_sum(scores(plays(lines())));
}

function solution2(): int
{
    return iterator_sum(scores(corrected_plays(plays(lines()))));
}

echo "Part 1: " . solution1() . "\n";
echo "Part 2: " . solution2() . "\n";

File: 02/solution.py
def plays(lines):
    """Map code to plays:
        A, X - rock
        B, Y - paper
        C, Z - scissors
    """
    for line in lines:
        opponent, you = line.split()
        yield ord(opponent) - ord("A"), ord(you) - ord("X")


def corrected_plays(plays):
    """The second value means lose, draw or win not rock, paper or scissors.
        0 (X) - lose
        1 (Y) - tie
        2 (Z) - win
    """
    # the second value means lose, draw or win not rock, paper, scizzors
    for opponent, you in plays:
        yield opponent, (opponent + you - 1) % 3


def scores(plays):
    for opponent, you in plays:
        yield 1 + you + ((you - opponent + 1) % 3) * 3


def solution1(lines):
    return sum(scores(plays(lines)))


def solution2(lines):
    return sum(scores(corrected_plays(plays(lines))))


if __name__ == "__main__":
    with open("input.txt") as f:
        print(f"Part 1: {solution1(f)}")
    with open("input.txt") as f:
        print(f"Part 2: {solution2(f)}")
